cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_resources.go

```diff
@@ -399,3 +406,117 @@ func (s *Server) updatePkgInstall(ctx context.Context, cluster, namespace string
+
+	return appLabelValue, nil
+}
+
+// findMatchingK8sResources returns the list of k8s resources matching the given listOptions
+// Code inspired by https://github.com/kubernetes/kubectl/blob/release-1.22/pkg/cmd/apiresources/apiresources.go#L142
+func (s *Server) findMatchingK8sResources(ctx context.Context, cluster string, listOptions metav1.ListOptions) ([]*corev1.ResourceRef, error) {
+	_, dynamicClient, discoveryClient, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get the k8s client: '%v'", err)
+	}
+
+	refs := []*corev1.ResourceRef{}
+
+	// fetch every possible kubernetes resource list that is available in the cluster
+	apiResourceLists, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return nil, err
+	}
+
+	// create channel to collect the matching k8s resources thus avoid blocking the waitgroup
+	// ussing a buffered channel with size 100; assumption: 60% native k8s resources + 40 CRDs
+	ch := make(chan resourceAndGvk, 100)
+	wg := sync.WaitGroup{}
+
+	// for each kubernetes resource list, filter out those not being having the verb "list"
+	for _, apiResourceList := range apiResourceLists {
+		if len(apiResourceList.APIResources) == 0 {
+			continue
+		}
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			continue
+		}
+
+		// for each kubernetes resource, spwan a goroutine to check if it matches the listOptions
+		for _, resource := range apiResourceList.APIResources {
+			if len(resource.Verbs) == 0 {
+				continue
+			}
+			// filter to resources that support the specified "List" verb
+			if !sets.NewString(resource.Verbs...).HasAll([]string{"list"}...) {
+				continue
+			}
+			wg.Add(1)
+			// spawn a goroutine to fetch the matching k8s resources
+			go func(resource metav1.APIResource, ch chan resourceAndGvk, wg *sync.WaitGroup) {
+				// convert the gvk to a gvr
+				gvk := gv.WithKind(resource.Kind)
+				gvr, _, err := s.kindToResource(s.restMapper, gvk)
+				if err != nil {
+					ch <- resourceAndGvk{nil, err}
+				}
+
+				// ignore the namespace as the resources may appear in different ns from which the PackageInstall is created
+				resources, err := dynamicClient.Resource(gvr).List(ctx, listOptions)
+				if err != nil {
+					ch <- resourceAndGvk{nil, err}
+				}
+
+				// for each found matching k8s resource, add it to the channel
+				for _, resource := range resources.Items {
+					ch <- resourceAndGvk{resourceRef: &corev1.ResourceRef{
+						ApiVersion: gvk.GroupVersion().String(),
+						Kind:       gvk.Kind,
+						Name:       resource.GetName(),
+						Namespace:  resource.GetNamespace(),
+					}}
+				}
+				wg.Done()
+			}(resource, ch, &wg)
+		}
+
+	}
+	// wait every goroutine to finish and close the channel
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	// once closed, iterate over the closed channel and push the objects to the response
```

Comment:
I don't think this comment is correct? That is, the channel will be closed when the `wg.Wait()` returns, but that's correctly happening inside a go routine. So the code below will start processing the data on the channel as it arrives (as it should) and will only complete (the `range ch`) when the channel is closed and all data has been drained.

Which still leaves me wondering why you need a large buffer on your channel? The work of the consumer here should be very quick compared to the producer (which is doing network fetches), so I'd not think you'd need a buffer at all.

Reply:
Ah, right! I mixed it up...  initially, I had an unbuffered channel, so I needed the goroutine encapsulating the `wg.wait()` and `close(ch)`. Then I moved to a buffered one, but left this goroutine back.

Yes.. the consumer is just aggregating data, which should be really fast; not really needing a buffer here.